pkg/component/etcd/etcd/crd.go

```diff
@@ -9,98 +9,146 @@ import (
 	_ "embed"
 	"fmt"
 
-	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/Masterminds/semver/v3"
+	druidcorev1alpha1 "github.com/gardener/etcd-druid/api/core/v1alpha1"
+	druidcorecrds "github.com/gardener/etcd-druid/api/core/v1alpha1/crds"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	"github.com/gardener/gardener/pkg/client/kubernetes"
 	"github.com/gardener/gardener/pkg/component"
 	"github.com/gardener/gardener/pkg/utils/flow"
 	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
-)
-
-var (
-	//go:embed crds/templates/crd-druid.gardener.cloud_etcds.yaml
-	// CRD holds the etcd custom resource definition template
-	CRD string
-	//go:embed crds/templates/crd-druid.gardener.cloud_etcdcopybackupstasks.yaml
-	etcdCopyBackupsTaskCRD string
-
-	etcdCRDName                = "etcds.druid.gardener.cloud"
-	etcdCopyBackupsTaskCRDName = "etcdcopybackupstasks.druid.gardener.cloud"
-	crdResources               []string
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
 )
 
 type crd struct {
-	client  client.Client
-	applier kubernetes.Applier
+	client    client.Client
+	crdGetter CRDGetter
 }
 
-func init() {
-	crdResources = append(crdResources, CRD, etcdCopyBackupsTaskCRD)
-}
-
-// NewCRD can be used to deploy the CRD definitions for Etcd and EtcdCopyBackupsTask.
-func NewCRD(c client.Client, applier kubernetes.Applier) component.Deployer {
-	return &crd{
-		client:  c,
-		applier: applier,
+// NewCRD can be used to deploy and/or retrieve the CRD definitions for all CRDs defined by etcd-druid.
+func NewCRD(c client.Client, k8sVersion *semver.Version) (component.Deployer, error) {
+	crdg, err := NewCRDGetter(k8sVersion)
+	if err != nil {
+		return nil, err
 	}
+	return &crd{
+		client:    c,
+		crdGetter: crdg,
+	}, nil
 }
 
+var _ component.Deployer = (*crd)(nil)
+
 // Deploy creates and updates the CRD definitions for Etcd and EtcdCopyBackupsTask.
 func (c *crd) Deploy(ctx context.Context) error {
 	var fns []flow.TaskFn
 
-	for _, resource := range crdResources {
-		r := resource
+	for _, resource := range c.crdGetter.GetAllCRDs() {
+		r := resource.DeepCopy()
 		fns = append(fns, func(ctx context.Context) error {
-			return c.applier.ApplyManifest(ctx, kubernetes.NewManifestReader([]byte(r)), kubernetes.DefaultMergeFuncs)
+			_, err := controllerutil.CreateOrPatch(ctx, c.client, r, nil)
```

Comment:
I think this is wrong usage of the func. `MutateFn` is nil, which means that if the CRD definition is changed, it won't update existing CRD resources in the cluster.
You can check the implementation of `CreateOrPatch`. If the resource exists, it relies on calling  `MutateFn` and computes the patch between the resource fetched from the apiserver and the object mutated by `MutateFn`.

Reply:
The reason i passed nil, is that the CRD definition cannot change for a specific version (released) of etcd-druid API. If it changes then it will have to be released and the new release version has to be used in g/g by explicitly changing the go.mod.

Reply:
I am not able to understand the reply. Do you say that you will do any API change in a new version of the CRD? Even, if. that's the case, the CRD versions are part of the CRD spec. Hence, the call here should be to update the existing CRDs in the system (update the CRD metadata and/or spec).

Reply:
Changes made as part of https://github.com/gardener/gardener/pull/11545/commits/e469199967b1256227f4367254263ca7d2db08db